src/common/src/util/memcmp_encoding.rs

```diff
@@ -0,0 +1,558 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::{Buf, BufMut};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+use super::iter_util::ZipEqFast;
+use crate::array::serial_array::Serial;
+use crate::array::{ArrayImpl, DataChunk};
+use crate::error::Result;
+use crate::row::Row;
+use crate::types::{
+    DataType, Datum, NaiveDateTimeWrapper, NaiveDateWrapper, NaiveTimeWrapper, OrderedF32,
+    OrderedF64, ScalarImpl, ToDatumRef,
+};
+use crate::util::sort_util::{ColumnOrder, OrderType};
+
+// NULL > any non-NULL value by default
+const DEFAULT_NULL_TAG_NONE: u8 = 1;
+const DEFAULT_NULL_TAG_SOME: u8 = 0;
+
+pub(crate) fn serialize_datum(
+    datum: impl ToDatumRef,
+    order: OrderType,
+    serializer: &mut memcomparable::Serializer<impl BufMut>,
+) -> memcomparable::Result<()> {
+    serializer.set_reverse(order.is_descending());
+    let (null_tag_none, null_tag_some) = if order.nulls_are_largest() {
+        (1u8, 0u8) // None > Some
+    } else {
+        (0u8, 1u8) // None < Some
+    };
+    if let Some(datum) = datum.to_datum_ref() {
+        null_tag_some.serialize(&mut *serializer)?;
+        datum.serialize(serializer)?;
+    } else {
+        null_tag_none.serialize(serializer)?;
+    }
+    Ok(())
+}
+
+pub(crate) fn serialize_datum_in_composite(
+    datum: impl ToDatumRef,
+    serializer: &mut memcomparable::Serializer<impl BufMut>,
+) -> memcomparable::Result<()> {
+    // NOTE: No need to call `serializer.set_reverse` because we are inside a
+    // composite type value, we should follow the outside order, except for `NULL`s.
+    if let Some(datum) = datum.to_datum_ref() {
+        DEFAULT_NULL_TAG_SOME.serialize(&mut *serializer)?;
```

Comment:
How can we know whether the inner datum should be NULLs largest or smallest? Or in Postgres we do not care about this?

Reply:
In PG, data items inside struct and array types are sorted with NULLs as largest regardless of outer order-by clause, which is the default behavior everywhere when there's no order-by clause.